File: Week 19/java/1.GreedyAlgorithms.java
/*
 * WEEK 19 - GREEDY ALGORITHMS
 * Topic: Greedy Paradigm + Classic Problems
 *
 * GREEDY STRATEGY:
 * At each step, make the locally optimal choice.
 * Hope that local optimum leads to global optimum.
 *
 * WHEN DOES GREEDY WORK?
 * The problem must have:
 * 1. GREEDY CHOICE PROPERTY: A global optimum can be reached by making locally optimal choices.
 * 2. OPTIMAL SUBSTRUCTURE: Optimal solution contains optimal solutions to subproblems.
 *
 * PROVING CORRECTNESS:
 * - Exchange argument: Show swapping greedy choice with any other choice doesn't improve the solution.
 * - Induction on greedy choices.
 *
 * PROBLEMS COVERED:
 * 1. Activity Selection (maximum non-overlapping intervals)
 * 2. Fractional Knapsack
 * 3. Minimum Number of Coins (canonical coin systems)
 * 4. Job Sequencing with Deadlines
 * 5. Merge Intervals
 *
 * IMPORTANT: Greedy DOESN'T always work! (e.g., 0/1 Knapsack needs DP)
 */

import java.util.*;

public class GreedyAlgorithms {

    // PROBLEM 1: Activity Selection
    // Given activities with start/end times, select maximum non-overlapping activities.
    // GREEDY: Always pick the activity that ENDS earliest.
    // Time: O(n log n) for sorting
    static int activitySelection(int[] start, int[] end) {
        int n = start.length;
        // Sort by end time
        Integer[] idx = new Integer[n];
        for (int i = 0; i < n; i++) idx[i] = i;
        Arrays.sort(idx, (a, b) -> end[a] - end[b]);

        int count = 1;
        int lastEnd = end[idx[0]];
        for (int i = 1; i < n; i++) {
            if (start[idx[i]] >= lastEnd) {
                count++;
                lastEnd = end[idx[i]];
            }
        }
        return count;
    }

    // PROBLEM 2: Fractional Knapsack
    // Can take fractions of items. Greedy: take by highest value/weight ratio.
    // Time: O(n log n)
    static double fractionalKnapsack(int[] weights, int[] values, int W) {
        int n = weights.length;
        Integer[] idx = new Integer[n];
        for (int i = 0; i < n; i++) idx[i] = i;
        // Sort by value/weight ratio (descending)
        Arrays.sort(idx, (a, b) -> Double.compare((double)values[b]/weights[b], (double)values[a]/weights[a]));

        double totalValue = 0;
        int remaining = W;
        for (int i : idx) {
            if (remaining <= 0) break;
            int take = Math.min(weights[i], remaining);
            totalValue += take * (double)values[i] / weights[i];
            remaining -= take;
        }
        return totalValue;
    }

    // PROBLEM 3: Minimum coins (standard greedy — works for canonical systems like USD)
    // NOTE: Greedy FAILS for non-canonical coin systems (e.g., {1,3,4} for target=6: greedy gives 4+1+1=3, DP gives 3+3=2)
    static Map<Integer, Integer> minCoinsGreedy(int[] coins, int target) {
        Arrays.sort(coins);
        Map<Integer, Integer> result = new LinkedHashMap<>();
        for (int i = coins.length - 1; i >= 0 && target > 0; i--) {
            if (coins[i] <= target) {
                int count = target / coins[i];
                result.put(coins[i], count);
                target -= coins[i] * count;
            }
        }
        return result;
    }

    // PROBLEM 4: Merge Intervals
    // Given list of intervals, merge all overlapping ones.
    // Time: O(n log n)
    static int[][] mergeIntervals(int[][] intervals) {
        if (intervals.length <= 1) return intervals;
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]); // sort by start
        List<int[]> merged = new ArrayList<>();
        int[] current = intervals[0];
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] <= current[1]) { // overlapping
                current[1] = Math.max(current[1], intervals[i][1]);
            } else {
                merged.add(current);
                current = intervals[i];
            }
        }
        merged.add(current);
        return merged.toArray(new int[0][]);
    }

    // PROBLEM 5: Minimum number of meeting rooms needed
    // Greedy + sorting + min-heap
    // Time: O(n log n)
    static int minMeetingRooms(int[][] intervals) {
        if (intervals.length == 0) return 0;
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        PriorityQueue<Integer> ends = new PriorityQueue<>(); // min-heap of end times
        ends.offer(intervals[0][1]);
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] >= ends.peek()) ends.poll(); // room becomes free
            ends.offer(intervals[i][1]);
        }
        return ends.size();
    }

    public static void main(String[] args) {
        // Activity Selection
        System.out.println("=== Activity Selection ===");
        int[] start = {1, 3, 0, 5, 8, 5};
        int[] end   = {2, 4, 6, 7, 9, 9};
        System.out.println("Max activities: " + activitySelection(start, end)); // 4

        // Fractional Knapsack
        System.out.println("\n=== Fractional Knapsack ===");
        System.out.printf("Max value: %.2f%n",
            fractionalKnapsack(new int[]{10,20,30}, new int[]{60,100,120}, 50)); // 240.0

        // Min Coins
        System.out.println("\n=== Min Coins (US denominations) ===");
        int[] usd = {1, 5, 10, 25, 50, 100};
        System.out.println("Change for 87 cents: " + minCoinsGreedy(usd, 87));
        System.out.println("Change for 30 cents: " + minCoinsGreedy(usd, 30));

        // Merge Intervals
        System.out.println("\n=== Merge Intervals ===");
        int[][] intervals = {{1,3},{2,6},{8,10},{15,18}};
        int[][] merged = mergeIntervals(intervals);
        System.out.print("Merged: ");
        for (int[] iv : merged) System.out.print(Arrays.toString(iv) + " ");
        System.out.println();
        // Expected: [1,6] [8,10] [15,18]

        // Meeting Rooms
        System.out.println("\n=== Min Meeting Rooms ===");
        int[][] meetings = {{0,30},{5,10},{15,20}};
        System.out.println("Min rooms needed: " + minMeetingRooms(meetings)); // 2
        int[][] meetings2 = {{7,10},{2,4}};
        System.out.println("Min rooms needed: " + minMeetingRooms(meetings2)); // 1
    }
}

File: Week 19/rust/1.GreedyAlgorithms.rs
/*
 * WEEK 19 - RUST DSA
 * Topic: Greedy Algorithms — Paradigm + Classic Problems
 * File: 1.GreedyAlgorithms.rs
 *
 * CONCEPT:
 *     Greedy algorithms commit to the locally optimal choice at each step.
 *     They work when the problem has greedy-choice property + optimal
 *     substructure. Typically O(n log n) due to a sort.
 *
 * KEY POINTS:
 *     - Sort + scan is the dominant pattern.
 *     - std::collections::BinaryHeap is a MAX-heap; wrap with Reverse for min-heap.
 *     - Use Vec::sort_by(|a,b| a.cmp(b)) or sort_by_key for ergonomic sorting.
 *     - Float comparators need partial_cmp + .unwrap() (NaN handling).
 *
 * ALGORITHM / APPROACH:
 *     Five canonical problems:
 *         1. Activity Selection         (sort by end time)
 *         2. Fractional Knapsack        (sort by value/weight desc)
 *         3. Min Coins (canonical)      (always take largest coin)
 *         4. Merge Intervals            (sort by start; merge overlaps)
 *         5. Min Meeting Rooms          (sort + min-heap of end times)
 *
 * RUST-SPECIFIC NOTES:
 *     - Use BinaryHeap<Reverse<i32>> for a min-heap. Reverse comes from
 *       std::cmp::Reverse.
 *     - Sorting indices: build (0..n).collect::<Vec<_>>() then sort_by_key.
 *     - Float compare: a.partial_cmp(&b).unwrap() (input must be finite).
 *     - Vec::iter().enumerate() is the idiomatic indexed loop.
 *
 * DRY RUN:
 *     Activity Selection
 *         start = [1,3,0,5,8,5]
 *         end   = [2,4,6,7,9,9]
 *         indices sorted by end: [0,1,3,2,4,5]
 *         pick 0 (1..2); 1 (3..4); 3 (5..7); skip 2 (0<7); pick 4 (8..9); skip 5.
 *         Answer = 4.
 *
 *     Min Meeting Rooms [[0,30],[5,10],[15,20]]
 *         heap=[Reverse(30)]; (5,10): 5<30 push 10; heap=[10,30].
 *         (15,20): 15>=10 pop; push 20; heap=[20,30]. size=2.
 *
 * COMPLEXITY:
 *     activity_selection      O(n log n)
 *     fractional_knapsack     O(n log n)
 *     min_coins_greedy        O(c log c)
 *     merge_intervals         O(n log n)
 *     min_meeting_rooms       O(n log n)
 */

use std::cmp::Reverse;
use std::collections::BinaryHeap;

// 1. ACTIVITY SELECTION
pub fn activity_selection(start: &[i32], end: &[i32]) -> i32 {
    let n = start.len();
    if n == 0 { return 0; }
    let mut idx: Vec<usize> = (0..n).collect();
    idx.sort_by_key(|&i| end[i]);
    let mut count = 1;
    let mut last_end = end[idx[0]];
    for &k in &idx[1..] {
        if start[k] >= last_end {
            count += 1;
            last_end = end[k];
        }
    }
    count
}

// 2. FRACTIONAL KNAPSACK
pub fn fractional_knapsack(weights: &[i32], values: &[i32], capacity: i32) -> f64 {
    let n = weights.len();
    let mut idx: Vec<usize> = (0..n).collect();
    idx.sort_by(|&a, &b| {
        let ra = values[b] as f64 / weights[b] as f64;
        let rb = values[a] as f64 / weights[a] as f64;
        ra.partial_cmp(&rb).unwrap()                    // descending density
    });
    let mut total = 0.0f64;
    let mut remaining = capacity;
    for i in idx {
        if remaining <= 0 { break; }
        let take = weights[i].min(remaining);
        total += take as f64 * values[i] as f64 / weights[i] as f64;
        remaining -= take;
    }
    total
}

// 3. MIN COINS (canonical systems)
pub fn min_coins_greedy(mut coins: Vec<i32>, mut target: i32) -> Vec<(i32, i32)> {
    coins.sort_by(|a, b| b.cmp(a));                     // descending
    let mut result = Vec::new();
    for c in coins {
        if target <= 0 { break; }
        if c <= target {
            let count = target / c;
            result.push((c, count));
            target -= c * count;
        }
    }
    result
}

// 4. MERGE INTERVALS
pub fn merge_intervals(mut intervals: Vec<(i32, i32)>) -> Vec<(i32, i32)> {
    if intervals.len() <= 1 { return intervals; }
    intervals.sort_by_key(|iv| iv.0);
    let mut merged: Vec<(i32, i32)> = Vec::with_capacity(intervals.len());
    merged.push(intervals[0]);
    for iv in intervals.into_iter().skip(1) {
        let last = merged.last_mut().unwrap();
        if iv.0 <= last.1 {
            last.1 = last.1.max(iv.1);
        } else {
            merged.push(iv);
        }
    }
    merged
}

// 5. MIN MEETING ROOMS
pub fn min_meeting_rooms(mut intervals: Vec<(i32, i32)>) -> i32 {
    if intervals.is_empty() { return 0; }
    intervals.sort_by_key(|iv| iv.0);
    let mut ends: BinaryHeap<Reverse<i32>> = BinaryHeap::new();
    ends.push(Reverse(intervals[0].1));
    for iv in intervals.into_iter().skip(1) {
        if let Some(&Reverse(top)) = ends.peek() {
            if iv.0 >= top {
                ends.pop();
            }
        }
        ends.push(Reverse(iv.1));
    }
    ends.len() as i32
}

fn main() {
    println!("=== Activity Selection ===");
    let start = [1, 3, 0, 5, 8, 5];
    let end   = [2, 4, 6, 7, 9, 9];
    println!("Max activities: {}", activity_selection(&start, &end));         // 4

    println!("\n=== Fractional Knapsack ===");
    println!("Max value: {:.2}",
             fractional_knapsack(&[10, 20, 30], &[60, 100, 120], 50));        // 240.00

    println!("\n=== Min Coins (US denominations) ===");
    let usd = vec![1, 5, 10, 25, 50, 100];
    println!("Change for 87: {:?}", min_coins_greedy(usd.clone(), 87));
    println!("Change for 30: {:?}", min_coins_greedy(usd.clone(), 30));

    println!("\n=== Merge Intervals ===");
    let merged = merge_intervals(vec![(1,3),(2,6),(8,10),(15,18)]);
    println!("Merged: {:?}", merged);

    println!("\n=== Min Meeting Rooms ===");
    println!("Rooms needed: {}", min_meeting_rooms(vec![(0,30),(5,10),(15,20)])); // 2
    println!("Rooms needed: {}", min_meeting_rooms(vec![(7,10),(2,4)]));          // 1
}

/*
 * NOTES (Rust vs Java):
 *   - BinaryHeap is max-heap by default; wrap with Reverse(x) for a min-heap
 *     of x.  Java has both PriorityQueue (min-heap) and reverse-order constructor.
 *   - Float sort: f64 doesn't implement Ord (NaN), so we use partial_cmp.unwrap().
 *     Java's Double.compare handles NaN with a specific ordering.
 *   - sort_by_key takes a closure returning a Key: Ord. For descending order,
 *     either negate or use sort_by with a flipped cmp.
 *   - Tuples (i32,i32) replace Java's int[2]; field access via .0, .1.
 *   - intervals.into_iter() consumes the Vec — efficient because we don't
 *     need the original any more.
 */
